### PR TITLE
Reshape Labels for Single-Task Datasets for torch.nn.CrossEntropyLoss

### PR DESCRIPTION
**Pull request by mufeili**
# Pull Request Template

## Description

As described in #2323 , `CrossEntropyLoss` expects the labels to have shape `(N)` for single-task cases, where `N` is the number of datapoints. We decide to reshape the labels for ~~the model wrappers.~~ the loss wrappers.

Can you review it? @nd-02110114 

`mypy` seems to yield errors irrelevant to this PR.

```bash
deepchem/utils/fragment_utils.py:194: error: Need type annotation for 'all_atoms' (hint: "all_atoms: List[<type>] = ...")
deepchem/data/datasets.py:2917: error: Need type annotation for 'datasets' (hint: "datasets: Dict[<type>, <type>] = ...")
deepchem/feat/molecule_featurizers/mol_graph_conv_featurizer.py:203: error: Need type annotation for 'src' (hint: "src: List[<type>] = ...")
deepchem/feat/molecule_featurizers/mol_graph_conv_featurizer.py:203: error: Need type annotation for 'dest' (hint: "dest: List[<type>] = ...")
deepchem/feat/molecule_featurizers/mol_graph_conv_featurizer.py:213: error: Need type annotation for 'bond_features' (hint: "bond_features: List[<type>] = ...")
deepchem/feat/graph_features.py:619: error: Need type annotation for 'next_adj' (hint: "next_adj: Set[<type>] = ...")
deepchem/data/data_loader.py:912: error: Need type annotation for 'remainder' (hint: "remainder: List[<type>] = ...")
deepchem/splits/splitters.py:563: error: Need type annotation for 'assigned' (hint: "assigned: Set[<type>] = ...")
deepchem/splits/splitters.py:1262: error: Need type annotation for 'scaffolds' (hint: "scaffolds: Dict[<type>, <type>] = ...")
deepchem/splits/splitters.py:1461: error: Need type annotation for 'years' (hint: "years: Dict[<type>, <type>] = ...")
deepchem/models/torch_models/torch_model.py:200: error: Need type annotation for '_prediction_outputs' (hint: "_prediction_outputs: List[<type>] = ...")
deepchem/models/torch_models/torch_model.py:201: error: Need type annotation for '_loss_outputs' (hint: "_loss_outputs: List[<type>] = ...")
deepchem/models/torch_models/torch_model.py:202: error: Need type annotation for '_variance_outputs' (hint: "_variance_outputs: List[<type>] = ...")
deepchem/models/torch_models/torch_model.py:203: error: Need type annotation for '_other_outputs' (hint: "_other_outputs: List[<type>] = ...")
deepchem/models/torch_models/torch_model.py:542: error: Need type annotation for 'access_values' (hint: "access_values: List[<type>] = ...")
deepchem/models/keras_model.py:208: error: Need type annotation for '_prediction_outputs' (hint: "_prediction_outputs: List[<type>] = ...")
deepchem/models/keras_model.py:209: error: Need type annotation for '_loss_outputs' (hint: "_loss_outputs: List[<type>] = ...")
deepchem/models/keras_model.py:210: error: Need type annotation for '_variance_outputs' (hint: "_variance_outputs: List[<type>] = ...")
deepchem/models/keras_model.py:211: error: Need type annotation for '_other_outputs' (hint: "_other_outputs: List[<type>] = ...")
deepchem/models/keras_model.py:627: error: Need type annotation for 'access_values' (hint: "access_values: List[<type>] = ...")
deepchem/dock/pose_generation.py:274: error: Need type annotation for 'docked_complexes' (hint: "docked_complexes: List[<type>] = ...")
deepchem/dock/pose_generation.py:275: error: Need type annotation for 'all_scores' (hint: "all_scores: List[<type>] = ...")
Found 22 errors in 9 files (checked 299 source files)
```

## Type of change

Please check the option that is related to your PR.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [ ] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [x] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.22.0**)
  - [ ] Run `mypy -p deepchem` and check no errors
  - [ ] Run `flake8 <modified file> --count` and check no errors
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

**Comment by peastman** [comment]
I think it would be better to make the change in `SparseSoftmaxCrossEntropy`.  Here is the documented behavior of that class:

>   The labels should have shape (batch_size) or (batch_size, tasks), and be
  integer class labels.  The outputs have shape (batch_size, classes) or
  (batch_size, tasks, classes) and be logits that are converted to probabilities
  using a softmax function.

If it isn't behaving that way, it's a bug.  Since it's implemented using `torch.nn.CrossEntropyLoss`, it needs to reshape its inputs as necessary to satisfy the needs of that class.  Something like this:

```python
if len(outputs.shape) == len(labels.shape) and labels.shape[1] == 1:
  labels = labels.squeeze(1)
```

**Comment by nd-02110114** [comment]
The mypy errors you shows doesn't happen in the CI and I also can't reproduce. 

> If it isn't behaving that way, it's a bug.

I  think the loss function behaves correctly.
The problems is that user often use the labels have shape (batch_size, 1) although the labels must have shape (batch_size,) for a single task classification. 

Actually, all molnet functions return the labels have shape (N, 1) for a single task classification. So, whenever some users use Models which have `SparseSoftmaxCrossEntropy ` with molnet functions, they need to reshape the labels. Mufei point out it is troblesome.

**Comment by peastman** [comment]
>  The problem is that user often use the labels have shape (batch_size, 1) although the labels must have shape (batch_size,) for a single task classification.

According to the docstring, (batch_size, 1) is also a legal shape for the input.  Also, the TensorFlow implementation works with inputs of that size.  It's just the PyTorch version that doesn't, and that's a bug.

**Comment by nd-02110114** [comment]
From the docstring, I assumed the loss function supports one to one relationship like the following.

labels `(batch_size)` <-> outputs `(batch_size, classes)`
labels `(batch_size, tasks)` <-> outputs `(batch_size, tasks, classes)`

But, there are no specific description about one to one relationship above. So, I could understand why you think it is bug. I'm not native in English, so I'm so sorry for your troublesome.

> Also, the TensorFlow implementation works with inputs of that size. It's just the PyTorch version that doesn't, and that's a bug.

OK. So, we should modify the loss functions directly and add more test cases.

**Comment by peastman** [comment]
> From the docstring, I assumed the loss function supports one to one relationship like the following.

Well, that wasn't what I intended when I wrote it. 😀  And that would certainly be inconvenient since, as you pointed out, the molnet loaders produce labels of shape (samples, 1).

**Comment by mufeili** [comment]
@peastman @nd-02110114 I've made the update for the loss wrapper.

**Review by rbharath — COMMENTED** [review]
Just had a chance to look at this PR. Changes look reasonable! Will defer to @peastman @nd-02110114 on when this is ready to merge in :)

File: deepchem/models/tests/test_losses.py [hunk 1]
```diff
@@ -198,6 +198,12 @@ def test_sparse_softmax_cross_entropy_pytorch(self):
     expected = [-np.log(softmax[0, 1]), -np.log(softmax[1, 0])]
     assert np.allclose(expected, result)
 
+    labels = torch.tensor([[1], [0]])
```
Comment (nd-02110114) [inline deepchem/models/tests/test_losses.py:201]: Can you add the same test for TensorFlow?

**Review by nd-02110114 — COMMENTED** [review]
I reviewed! Basically, it looks good to me.

File: deepchem/models/tests/test_losses.py (hunk 1 above)
Comment (mufeili) [reply deepchem/models/tests/test_losses.py:207]: Done.

**Comment by peastman** [comment]
The tests don't have any assertions.  Are they just making sure it doesn't crash?

**Comment by mufeili** [comment]
> The tests don't have any assertions. Are they just making sure it doesn't crash?

Yes.

**Comment by peastman** [comment]
Ok.  Looks like you have a test failure in `test_sparse_softmax_cross_entropy_tf()`.

**Comment by mufeili** [comment]
> Ok. Looks like you have a test failure in `test_sparse_softmax_cross_entropy_tf()`.

This should be fixed now.

File: deepchem/models/losses.py [hunk 2]
```diff
@@ -187,7 +187,12 @@ class SparseSoftmaxCrossEntropy(Loss):
 
   def _compute_tf_loss(self, output, labels):
     import tensorflow as tf
+
+    if labels.shape[-1] == 1:
+      labels = tf.squeeze(labels, axis=-1)
```
Comment (peastman) [inline deepchem/models/losses.py:192]: I don't think that's the right check for TensorFlow.  `sparse_softmax_cross_entropy_with_logits()` requires `labels` to have one fewer dimension than `output`.  We should check whether `len(labels.shape) == len(output.shape)+1`.

Comment (mufeili) [reply deepchem/models/losses.py:192]: How about adding an additional assertion statement after L192 for the check you proposed? There can be multiple reasons for this. L191/192 are mainly handling the issue in molnet data loader.

Comment (peastman) [reply deepchem/models/losses.py:192]: It's not an assertion.  The issue is that this function is more permissive of its inputs than the TensorFlow and PyTorch functions we use to implement it.  Given any inputs that are legal for this function, we need to reshape them as required for the TensorFlow and PyTorch functions.  Conversely, we should only reshape them if the inputs are *not* the correct shape for the implementing function.  If the inputs already have the shape expected by `sparse_softmax_cross_entropy_with_logits()`, we shouldn't call `squeeze()`.

Comment (mufeili) [reply deepchem/models/losses.py:192]: What check/reshape will you suggest? It's not guaranteed that one can obtain `len(labels.shape) == len(output.shape)+1` by reshaping as the inputs may not be valid themselves.

Comment (peastman) [reply deepchem/models/losses.py:192]: If the inputs aren't valid, that's the user's problem, not ours. 😀  How about

```python
if len(labels.shape) == len(outputs.shape):
  labels = tf.squeeze(labels, axis=-1) 
```

Comment (mufeili) [reply deepchem/models/losses.py]: Sounds good. I've made the update.

Comment (mufeili) [reply deepchem/models/losses.py]: With "thumbs up", I assume it looks good to you 😄 . Any further issues? @peastman

Comment (peastman) [reply deepchem/models/losses.py]: I think it's good to merge.

**Comment by mufeili** [comment]
@rbharath You may go ahead and merge it. The failed tests are related to DTNN,
